discovery-provider/src/solana/anchor_program_indexer.py

```diff
@@ -74,23 +80,146 @@ def get_latest_slot(self):
+            db_models: Dict = defaultdict(lambda: defaultdict(lambda: []))
+            if entity_ids["users"]:
+                existing_users = (
+                    session.query(User)
+                    .filter(
+                        User.is_current,
+                        User.user_id.in_(list(entity_ids["users"])),
+                    )
+                    .all()
                 )
+                for user in existing_users:
+                    db_models["users"][user.user_id] = [user]
+
+            # TODO: Find all other track/playlist/etc. models
+
+            self.process_transactions(
+                session, processed_transactions, db_models, metadata_dictionary
+            )
+
+    def process_transactions(
+        self,
+        session: Session,
+        processed_transactions: List[ParsedTx],
+        db_models: Dict,
+        metadata_dictionary: Dict,
+    ):
+        records: List[Any] = []
+        for transaction in processed_transactions:
+            instructions = transaction["tx_metadata"]["instructions"]
+            for instruction in instructions:
+                instruction_name = instruction.get("instruction_name")
+                if instruction_name in transaction_handlers:
+                    transaction_handlers[instruction_name](
+                        session,
+                        transaction,
+                        instruction,
+                        db_models,
+                        metadata_dictionary,
+                        records,
+                    )
 
-    async def parse_tx(self, tx_sig):
+        self.invalidate_old_records(session, db_models)
+        session.bulk_save_objects(records)
+
+    def invalidate_old_records(self, session: Session, db_models: Dict[str, Dict]):
+        # Update existing record in db to is_current = False
+        if db_models.get("users"):
+            user_ids = list(db_models.get("users", {}).keys())
+            self.msg(f"{user_ids}")
+            session.query(User).filter(
+                User.is_current == True, User.user_id.in_(user_ids)
+            ).update({"is_current": False}, synchronize_session="fetch")
+
+    def extract_ids(self, processed_transactions: List[ParsedTx]):
+        entites: Dict[str, Set[str]] = defaultdict(lambda: set())
+
+        for transaction in processed_transactions:
+            instructions = transaction["tx_metadata"]["instructions"]
+            for instruction in instructions:
+                instruction_name = instruction.get("instruction_name")
+                if instruction_name == "init_admin":
+                    pass
+                elif instruction_name == "update_admin":
+                    pass
+                elif instruction_name == "init_user":
+                    # No action to be taken here
+                    pass
+
+                elif instruction_name == "init_user_sol":
+                    # TODO: parse the tx data for their user id and fetch
```

Comment:
also, i just made that change in instruction validation to check for accounts named admin. could we do a similar thing here? if ID is in the decoded data, extract it. maybe this could be in the anchor parser?

Reply:
Yea, I figured that this would be similar logic to both the metadata and validation steps.
Did you mean that all of this could be in the anchor parser?